File: src/buddy_cli/doctor.py
"""Buddy installation diagnostics."""

from __future__ import annotations

import os
from dataclasses import asdict, dataclass
from pathlib import Path
from typing import Dict, List, Optional

from buddy_cli.config import BuddyConfig, ConfigError, ConfigStore
from buddy_cli.constants import DEFAULT_MODEL, DEFAULT_OLLAMA_BASE_URL
from buddy_cli.ollama import OllamaClient, OllamaError
from buddy_cli.paths import AppPaths
from buddy_cli.runtime_manager import RuntimeManager
from buddy_cli.runtime_manifest import resolve_runtime_spec


@dataclass(frozen=True)
class DiagnosticCheck:
    status: str
    name: str
    detail: str


@dataclass(frozen=True)
class DiagnosticReport:
    checks: List[DiagnosticCheck]

    @property
    def healthy(self) -> bool:
        return not any(check.status == "FAIL" for check in self.checks)

    def to_dict(self) -> Dict[str, object]:
        return {
            "healthy": self.healthy,
            "checks": [asdict(check) for check in self.checks],
        }


class Doctor:
    """Inspect Buddy without changing system or application state."""

    def __init__(
        self,
        paths: AppPaths,
        config_store: ConfigStore,
        runtime_manager: RuntimeManager,
    ) -> None:
        self.paths = paths
        self.config_store = config_store
        self.runtime_manager = runtime_manager

    def run(self) -> DiagnosticReport:
        checks: List[DiagnosticCheck] = []
        try:
            spec = resolve_runtime_spec()
            checks.append(
                DiagnosticCheck(
                    "PASS",
                    "platform",
                    f"managed runtime available for {spec.operating_system}/{spec.architecture}",
                )
            )
        except Exception as exc:
            checks.append(DiagnosticCheck("FAIL", "platform", str(exc)))

        writable_target = self.paths.root
        while (
            not writable_target.exists() and writable_target != writable_target.parent
        ):
            writable_target = writable_target.parent
        if os.access(writable_target, os.W_OK):
            checks.append(
                DiagnosticCheck(
                    "PASS",
                    "storage",
                    f"Buddy can write under {self.paths.root}",
                )
            )
        else:
            checks.append(
                DiagnosticCheck(
                    "FAIL",
                    "storage",
                    f"Buddy cannot write under {self.paths.root}",
                )
            )

        config: Optional[BuddyConfig]
        try:
            config = self.config_store.load()
            if config:
                checks.append(
                    DiagnosticCheck(
                        "PASS",
                        "configuration",
                        f"configured for {config.provider} Ollama",
                    )
                )
            else:
                checks.append(
                    DiagnosticCheck(
                        "FAIL",
                        "configuration",
                        "Buddy has not been set up; run 'buddy setup'",
                    )
                )
        except ConfigError as exc:
            config = None
            checks.append(DiagnosticCheck("FAIL", "configuration", str(exc)))

        if config and config.provider == "managed":
            executable = Path(config.executable) if config.executable else None
            if executable and executable.is_file():
                checks.append(DiagnosticCheck("PASS", "runtime", str(executable)))
            else:
                checks.append(
                    DiagnosticCheck(
                        "FAIL", "runtime", "managed Ollama executable is missing"
                    )
                )
        else:
            executable = self.runtime_manager.find_system_executable()
            checks.append(
                DiagnosticCheck(
                    "PASS" if executable else "WARN",
                    "runtime",
                    str(executable)
                    if executable
                    else "system Ollama executable not found",
                )
            )

        base_url = config.base_url if config else DEFAULT_OLLAMA_BASE_URL
        model = config.model if config else DEFAULT_MODEL
        try:
            client = OllamaClient(base_url, timeout=2.0)
            version = client.get_version()
            checks.append(
                DiagnosticCheck(
                    "PASS", "api", f"Ollama {version.version} at {base_url}"
                )
            )
            if client.has_model(model):
                checks.append(DiagnosticCheck("PASS", "model", model))
            else:
                checks.append(
                    DiagnosticCheck(
                        "FAIL", "model", f"{model} is not installed; run 'buddy setup'"
                    )
                )
        except (OllamaError, ValueError) as exc:
            checks.append(DiagnosticCheck("FAIL", "api", str(exc)))
            checks.append(
                DiagnosticCheck(
                    "WARN",
                    "model",
                    "model could not be checked while Ollama is offline",
                )
            )

        return DiagnosticReport(checks)

File: src/doctor.rs
//! Buddy installation diagnostics.

use serde::{Deserialize, Serialize};
use std::fs::{self, OpenOptions};
use std::path::PathBuf;
use std::time::Duration;

use crate::config::{BuddyConfig, ConfigStore};
use crate::constants::{DEFAULT_MODEL, DEFAULT_OLLAMA_BASE_URL};
use crate::ollama::OllamaClient;
use crate::paths::AppPaths;
use crate::runtime_manager::RuntimeManager;
use crate::runtime_manifest::resolve_runtime_spec;

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DiagnosticCheck {
    pub status: String,
    pub name: String,
    pub detail: String,
}

#[derive(Debug, Clone, PartialEq, Eq, Serialize, Deserialize)]
pub struct DiagnosticReport {
    pub healthy: bool,
    pub checks: Vec<DiagnosticCheck>,
}

impl DiagnosticReport {
    pub fn new(checks: Vec<DiagnosticCheck>) -> Self {
        let healthy = !checks.iter().any(|c| c.status == "FAIL");
        Self { healthy, checks }
    }
}

pub struct Doctor {
    pub paths: AppPaths,
    pub config_store: ConfigStore,
    pub runtime_manager: RuntimeManager,
}

impl Doctor {
    pub fn new(
        paths: AppPaths,
        config_store: ConfigStore,
        runtime_manager: RuntimeManager,
    ) -> Self {
        Self {
            paths,
            config_store,
            runtime_manager,
        }
    }

    fn check_storage_writable(&self) -> bool {
        let mut target = self.paths.root.clone();
        while !target.exists() {
            if let Some(parent) = target.parent() {
                if parent == target {
                    break;
                }
                target = parent.to_path_buf();
            } else {
                break;
            }
        }

        let probe = target.join(".buddy-probe-write-test");
        if let Ok(mut file) = OpenOptions::new()
            .write(true)
            .create(true)
            .truncate(true)
            .open(&probe)
        {
            use std::io::Write;
            let _ = file.write_all(b"probe");
            drop(file);
            let _ = fs::remove_file(&probe);
            true
        } else {
            false
        }
    }

    pub fn run(&self) -> DiagnosticReport {
        let mut checks = Vec::new();

        // 1. Platform check
        match resolve_runtime_spec() {
            Ok(spec) => {
                checks.push(DiagnosticCheck {
                    status: "PASS".to_string(),
                    name: "platform".to_string(),
                    detail: format!(
                        "managed runtime available for {}/{}",
                        spec.operating_system, spec.architecture
                    ),
                });
            }
            Err(e) => {
                checks.push(DiagnosticCheck {
                    status: "FAIL".to_string(),
                    name: "platform".to_string(),
                    detail: e.to_string(),
                });
            }
        }

        // 2. Storage check
        if self.check_storage_writable() {
            checks.push(DiagnosticCheck {
                status: "PASS".to_string(),
                name: "storage".to_string(),
                detail: format!("Buddy can write under {}", self.paths.root.display()),
            });
        } else {
            checks.push(DiagnosticCheck {
                status: "FAIL".to_string(),
                name: "storage".to_string(),
                detail: format!("Buddy cannot write under {}", self.paths.root.display()),
            });
        }

        // 3. Configuration check
        let config: Option<BuddyConfig> = match self.config_store.load() {
            Ok(Some(cfg)) => {
                checks.push(DiagnosticCheck {
                    status: "PASS".to_string(),
                    name: "configuration".to_string(),
                    detail: format!("configured for {} Ollama", cfg.provider),
                });
                Some(cfg)
            }
            Ok(None) => {
                checks.push(DiagnosticCheck {
                    status: "FAIL".to_string(),
                    name: "configuration".to_string(),
                    detail: "Buddy has not been set up; run 'buddy setup'".to_string(),
                });
                None
            }
            Err(e) => {
                checks.push(DiagnosticCheck {
                    status: "FAIL".to_string(),
                    name: "configuration".to_string(),
                    detail: e.to_string(),
                });
                None
            }
        };

        // 4. Runtime check
        if let Some(ref cfg) = config {
            if cfg.provider == "managed" {
                let exe = cfg.executable.as_ref().map(PathBuf::from);
                if let Some(ref path) = exe {
                    if path.is_file() {
                        checks.push(DiagnosticCheck {
                            status: "PASS".to_string(),
                            name: "runtime".to_string(),
                            detail: path.display().to_string(),
                        });
                    } else {
                        checks.push(DiagnosticCheck {
                            status: "FAIL".to_string(),
                            name: "runtime".to_string(),
                            detail: "managed Ollama executable is missing".to_string(),
                        });
                    }
                } else {
                    checks.push(DiagnosticCheck {
                        status: "FAIL".to_string(),
                        name: "runtime".to_string(),
                        detail: "managed Ollama executable is missing".to_string(),
                    });
                }
            } else {
                let exe = RuntimeManager::find_system_executable();
                if let Some(path) = exe {
                    checks.push(DiagnosticCheck {
                        status: "PASS".to_string(),
                        name: "runtime".to_string(),
                        detail: path.display().to_string(),
                    });
                } else {
                    checks.push(DiagnosticCheck {
                        status: "WARN".to_string(),
                        name: "runtime".to_string(),
                        detail: "system Ollama executable not found".to_string(),
                    });
                }
            }
        } else {
            let exe = RuntimeManager::find_system_executable();
            if let Some(path) = exe {
                checks.push(DiagnosticCheck {
                    status: "PASS".to_string(),
                    name: "runtime".to_string(),
                    detail: path.display().to_string(),
                });
            } else {
                checks.push(DiagnosticCheck {
                    status: "WARN".to_string(),
                    name: "runtime".to_string(),
                    detail: "system Ollama executable not found".to_string(),
                });
            }
        }

        // 5. API & Model checks
        let base_url = config
            .as_ref()
            .map(|c| c.base_url.as_str())
            .unwrap_or(DEFAULT_OLLAMA_BASE_URL);
        let model = config
            .as_ref()
            .map(|c| c.model.as_str())
            .unwrap_or(DEFAULT_MODEL);

        match OllamaClient::with_timeout(base_url, Duration::from_secs(2)) {
            Ok(client) => match client.get_version() {
                Ok(version) => {
                    checks.push(DiagnosticCheck {
                        status: "PASS".to_string(),
                        name: "api".to_string(),
                        detail: format!("Ollama {} at {}", version.version, base_url),
                    });
                    if client.has_model(model) {
                        checks.push(DiagnosticCheck {
                            status: "PASS".to_string(),
                            name: "model".to_string(),
                            detail: model.to_string(),
                        });
                    } else {
                        checks.push(DiagnosticCheck {
                            status: "FAIL".to_string(),
                            name: "model".to_string(),
                            detail: format!("{} is not installed; run 'buddy setup'", model),
                        });
                    }
                }
                Err(e) => {
                    checks.push(DiagnosticCheck {
                        status: "FAIL".to_string(),
                        name: "api".to_string(),
                        detail: e.to_string(),
                    });
                    checks.push(DiagnosticCheck {
                        status: "WARN".to_string(),
                        name: "model".to_string(),
                        detail: "model could not be checked while Ollama is offline".to_string(),
                    });
                }
            },
            Err(e) => {
                checks.push(DiagnosticCheck {
                    status: "FAIL".to_string(),
                    name: "api".to_string(),
                    detail: e.to_string(),
                });
                checks.push(DiagnosticCheck {
                    status: "WARN".to_string(),
                    name: "model".to_string(),
                    detail: "model could not be checked while Ollama is offline".to_string(),
                });
            }
        }

        DiagnosticReport::new(checks)
    }
}
